### ISSUE TITLE
error: executing query failed: db error: ERROR: Evaluation error: internal error: Mismatched aggregates for key in ReduceCollation

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#27455]
### What version of Materialize are you using?

4c9c744a16

### What is the issue?

```
^^^ +++
27:1: error: executing query failed: db error: ERROR: Evaluation error: internal error: Mismatched aggregates for key in ReduceCollation: ERROR: Evaluation error: internal error: Mismatched aggregates for key in ReduceCollation
     |
   2 | $ postgres-execute c ... [rest of line truncated for security]
  26 | 
  27 | > SELECT f1, max(f2), SUM(LENGTH(f3)) FROM mysql_source_tableC_false GROUP BY f1;
     | ^
+++ !!! Error Report
1 errors were encountered during execution
source: /var/lib/buildkite-agent/builds/buildkite-large-aarch64-e5872a3-i-0c9bedb08bf81aea3-1/materialize/nightly/misc/python/materialize/checks/all_checks/mysql_cdc.py:242
```
Happened in tests toggling txn-wal: https://buildkite.com/materialize/nightly/builds/7972#018fe44e-978d-412c-b146-359b8ee600a5
Seems to be a flake, but happened in two tests in the same Nightly, both times with MySQL. CC @rjobanp

ci-regexp: Mismatched aggregates for key in ReduceCollation

**Comment by def-** [comment]
Also happened with Postgres source: https://buildkite.com/materialize/test/builds/83230#018fe8f2-517c-4dcf-b0a7-1162b2eede68

**Comment by jkosh44** [comment]
> Also happened with Postgres source: https://buildkite.com/materialize/test/builds/83230#018fe8f2-517c-4dcf-b0a7-1162b2eede68

The fact that we see the same exact error in test that doesn't touch the txn wal flag suggests to me that the failure is unrelated to txn wal. I think this should be re-assigned to compute.

**Comment by def-** [comment]
@petrosagg @rjobanp  I'm afraid it might be a regression and we might somehow have started getting corrupt data in sources. At first it was only MySQL + txn-wal, but now also Postgres and without txn-wal, so this seems to be a Storage-related issue.

**Comment by nrainer-materialize** [comment]
@petrosagg, @rjobanp: Do you agree that this should be considered as release blocker?

**Comment by rjobanp** [comment]
Yes I think we need to get to the bottom of it. But it's unclear whether this is really a storage issue or just a symptom downstream from storage sources. There haven't been any significant mysql/postgres source changes in the last week or so -- the only one that is potentially related is the reclock operator work. Is there any way to isolate a range of commits to look at?

**Comment by nrainer-materialize** [comment]
> Is there any way to isolate a range of commits to look at?

Unfortunately, that is difficult to determine for a flake.

All known occurrences are:
```
Match in build #8003 (pipeline nightly on main):
URL: https://buildkite.com/materialize/nightly/builds/8003
Annotation: Checks toggling persist batch columnar format
-------------------------------------------------------------------------------------
Match in build #7973 (pipeline nightly on aljoscha:compute-read-only-clusters):
URL: https://buildkite.com/materialize/nightly/builds/7973
Annotation: Checks + kill storage clusterd
-------------------------------------------------------------------------------------
Match in build #7972 (pipeline nightly on main):
URL: https://buildkite.com/materialize/nightly/builds/7972
Annotation: Checks + txn_wal fencing
-------------------------------------------------------------------------------------
Match in build #7972 (pipeline nightly on main):
URL: https://buildkite.com/materialize/nightly/builds/7972
Annotation: Checks + Toggle txn_wal
```

That is, it was earliest observed in [builds/7972](https://buildkite.com/materialize/nightly/builds/7972) this Tuesday on commit 4c9c744a167a8ee1bc10584e774a058df4575fd9. All we can tell from that is that the potential fault was introduced at or before this commit...

**Comment by def-** [comment]
I'll try to rerun platform checks on some older states before 4c9c744 in the hopes of getting a closer range of commits.

**Comment by def-** [comment]
Based on this the range seems to be https://github.com/MaterializeInc/materialize/compare/9565162583...29f53f0407 so the most likely responsible changes is https://github.com/MaterializeInc/materialize/pull/27058 (I guess https://github.com/MaterializeInc/materialize/pull/27376 is less risky)

**Comment by jkosh44** [comment]
This is resolved by #27502 (CC @petrosagg @rjobanp).
